rebuild-popular-crates.sh

```diff
@@ -1,25 +1,45 @@
 #!/bin/bash
 
+# pup can be installed via: go get github.com/ericchiang/pup
+# uq can be installed using cargo-quickinstall
+
+set -euo pipefail
+
+which pup || (echo "pup can be installed via: go get github.com/ericchiang/pup"; exit 1)
+which uq || (echo "uq can be installed using cargo-quickinstall"; exit 1)
+
 function get_top() {
-    curl https://lib.rs/$1 > ./temp.html
-    cat temp.html |
+    curl --fail "https://lib.rs/$1" |
         pup ':parent-of(:parent-of(:parent-of(.bin))) json{}' |
         jq -r '.[] |
             (.children[1].children|map(select(.class == "downloads").title)[0]// "0 ")
             + ":" +
             (.children[0].children[0].text)' |
         sort -gr |
         grep -v / |
+        grep -v ^0 |
         head -n 100 |
-        sed s/^.*:// |
-        sort
-    rm ./temp.html
+        tee /dev/fd/2 | # debugging goes to stderr
+        sed s/^.*://
+
+    echo "done with $1" 1>&2
+}
+
+function get_top_both() {
+    (
+        get_top command-line-utilities
+        get_top development-tools/cargo-plugins
+    ) | sort
+}
+
+function get_new_file_contents() {
+    (
+        cat popular-crates.txt | grep -B10000 '####################################'
+        get_top_both
+    ) | uq
 }
 
-top_crates=""
-top_crates="${top_crates} $(get_top command-line-utilities)"
-top_crates="${top_crates} $(get_top development-tools/cargo-plugins)"
-top_crates=`echo ${top_crates} | uniq | tr " " "\n" | sort`
```

Comment:
When I was looking at this, I don't think that this `uniq` ever did anything, because it is handed a single line with all of the crates on. `uniq` needs to be handed a sorted list of lines, and it will only deduplicate lines that are immediately after each other.

In order to make it work when we have blessed crates at the top of the file, I installed `uq`. This doesn't require the list to be sorted, and can do the right thing for `cargo-make`.

Reply:
thanks